EMP-304 dummy storage device

raysiudak opened the pull request.

raysiudak committed:
Some minor cosmetic changes

- Renamed 'packages' to 'commands'
- Print the found arm emulator for consistant stdout

Contributes to EMP-272 and EMP-304

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add test functionality for testing the prepare_disk.sh script.
This commit requires a dockerimage rebuild.

This includes:

- Test storage device generation, that as a bonus test the tools in the toolbox
- The first example test testing shrinking spool partition happy-day scenario
- Trigger CI for a new dockerimage build

Contributes to EMP-272, implements EMP-304

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r commented on tests/buildenv.sh:
```diff
@@ -5,8 +5,8 @@ set -eu
-FILESYSTEMS="overlay squashfs tmpfs"
-COMMANDS="apk mksquashfs xz"
+FILESYSTEMS="overlay squashfs tmpfs ext4 f2fs"
+COMMANDS="apk mksquashfs xz sed"
```
keep these variables sorted please

oliv3r commented on tests/buildenv.sh:
```diff
@@ -38,6 +38,7 @@ check_emulation_support()
 
         if grep -q "${ARMv7_MAGIC}" "${emu}"; then
             __ARM_EMU_BIN="$(sed 's/interpreter //;t;d' "${emu}")"
+            echo "${__ARM_EMU_BIN}"
```
this is printed too early. This is only the expected emulator. There are 3 tests following this one which can fail. resulting in a printout saying 'emu support: qemu-arm-static'  'Error, cannot continue'

Right now we print 'ok' whenever all checks have passed. I think it's not a bad idea to print more info on success to yield more information, but then it should be at the end, and in this specific case using ${ARM_EMU_BIN}.

so pick one:
```sh
echo "ok"
```
(that's what we have now)

or
```sh
echo "ok (${QEMU_ARM_BIN)"
```
or something like that.
```

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,19 @@
+
+# Test storage device parameters
+STORAGE_DEVICE_FILE="/tmp/storage_device.img"
+SECTOR_SIZE="512"
```
:+1:

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,19 @@
+# Test storage device parameters
+STORAGE_DEVICE_FILE="/tmp/storage_device.img"
+SECTOR_SIZE="512"
+SECTOR_TOTAL="7553024"
```
DEVICE_TOTAL_SIZE (maybe add a comment 'Storage size in sectors, real size xxx GiB)?

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,19 @@
+SECTOR_TOTAL="7553024"
+BOOT_START="2048"
+ROOTFS_START="67584"
+SPOOL_START="1998848"
```
it is actually the USERDATA_START. We just mount the cluster spool there because we didn't have time to properly implement it, so lets not let this technical debt accumulate further.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+    device_file="${3}"
+
+    loop_device="$(chroot "${rootfs_dir}" /sbin/losetup -f)"
+    echo "Using next available loop device '${loop_device}' to format partition using: '${command}':"
```
i think you intended a period at the end of this line?

Ah I think it prints the mount point of the if grep at the end here? If so, i think you need an extra space :) (or maybe even ''${command}' on: "?)

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+    echo "Using next available loop device '${loop_device}' to format partition using: '${command}':"
+
+    touch "${rootfs_dir}${loop_device}"
+    mount --bind "${loop_device}" "${rootfs_dir}${loop_device}"
```
i thought you said you wanted to do this during setup/teardown? if you do it in the setup/teardown you can just dd the file INTO the '${rootfsdir}', unless of course you want the image the remain the same between setup-teardowns of course, which makes sense here, as you prepare the image for the last test; though then it does make sense to remove the earlier instance of the test image that is used during setup/teardown :)

raysiudak replied:
create_dummy_storage_device is called in setup and the images file is destroyed in teardown

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+    sector_start="${2}"
+    device_file="${3}"
+
+    loop_device="$(chroot "${rootfs_dir}" /sbin/losetup -f)"
```
if we chroot, do we not have access to the PATH? I think it's better not to prefix the commands if we don't have to, so that if they'd move, things don't break

raysiudak replied:
Nope

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+    if grep "${loop_device}" "/proc/mounts"; then
+        umount "${rootfs_dir}${loop_device}"
+        unlink "${rootfs_dir}${loop_device}"
+    fi
```
For concistency with teardown:
```sh
        if [ -f "${loop_device}" ]; then
            unlink "${loop_device}"
        fi
```
so that it uses the same pattern

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+{
+    partition_table_file="${rootfs_dir}/tmp/partition_table"
+
+cat <<EOT >> "${partition_table_file}"
```
you can, if you want, indent cat by one as you expect, and if you use <<-EOT (added hyphen) you can double indent the partition table, as the hyphen removes the indentation in the temporary file) (sadly however, this does not apply to the EOT marker as far as i'm aware.

raysiudak replied:
Because EOT end should start line column 1, i think it makes it more readable if the start is also on the first line.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+spool   : start=${SPOOL_START},     size=$(( SECTOR_TOTAL - SPOOL_START )),   Id=83
+EOT
+
+    echo "Creating test image: '${rootfs_dir}${STORAGE_DEVICE_FILE}'"
```
When you start with a capital, end with a period :p

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+cat <<EOT >> "${partition_table_file}"
+unit: sectors
+
+boot    : start=${BOOT_START},      size=$(( ROOTFS_START - BOOT_START )),    Id=83
```
we only space around curly braces when we initialize arrays

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+
+    echo "Writing partition table:"
+    cat "${partition_table_file}"
+    chroot "${rootfs_dir}" /sbin/sfdisk "${STORAGE_DEVICE_FILE}" < "${partition_table_file}" || return 1
```
as with losetup, drop the path if you can

raysiudak replied:
No cannot because we have no environment when running the command.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+
+    echo "Creating test image: '${rootfs_dir}${STORAGE_DEVICE_FILE}'"
+    dd if="/dev/zero" of="${rootfs_dir}${STORAGE_DEVICE_FILE}" bs="1" count="0" \
+        seek="$(( SECTOR_SIZE * SECTOR_TOTAL ))" || return 1
```
if dd fails, it will return with 1 already, no need to explicitly return on failure

raysiudak replied:
Yeah, forgot to remove this. Before moving this code to the setup, i wanted the function to return control instead of stopping the script, so that it was possible to print the failure exit text.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+{
+    command="${1}"
+    sector_start="${2}"
+    device_file="${3}"
```
device_file == STORAGE_DEVICE_FILE? do we forsee (YAGNI) that we need this parameter variable? It's a 'constant' throughout the program already anyway ...

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+    format_dummy_partition "/usr/sbin/mkfs.f2fs" "$(( SECTOR_SIZE * ROOTFS_START ))" "${STORAGE_DEVICE_FILE}" || return 1
+    format_dummy_partition "/usr/sbin/mkfs.f2fs" "$(( SECTOR_SIZE * SPOOL_START ))" "${STORAGE_DEVICE_FILE}" || return 1
+
+    unlink "${partition_table_file}"
```
because you create this in this function, i understand why you clean it up at the end again. However if any of the previous commands fail, this will never get cleaned.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +24,55 @@ result=0
+
+create_dummy_storage_device()
+{
+    partition_table_file="${rootfs_dir}/tmp/partition_table"
```
why store this in a file, if it is created AND removed in this function only. Just store the data in a variable and use the variable as stdin if possible. Unless it is really needed avoid storing temporary data in files

raysiudak replied:
See comment down the line, i don't see how it is possible using chroot

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -51,6 +113,10 @@ teardown()
         return
     fi
 
+    if [ -r "${STORAGE_DEVICE_FILE}" ]; then
```
during cleanup, i think we can use -e or -f; the file does not have to be readable by the user. I don't even think we need to have write access to the file, we just need write access to the upper laying directory

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -51,6 +113,10 @@ teardown()
     fi
 
+    if [ -r "${STORAGE_DEVICE_FILE}" ]; then
+        unlink "${STORAGE_DEVICE_FILE}"
```
could it be the case that it is still mounted? if not, then this is obviously fine

raysiudak replied:
Still being mounted would be an error in the setup and then the script should stop

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -81,6 +147,9 @@ failure_exit()
     echo "    umount '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
     echo "    unlink '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
+    echo "    for loop_device in /dev/loop[0-9]; do ( if grep '${loop_device}' '/proc/mounts'; \\ "
+    echo "    then sudo umount '${rootfs_dir}${loop_device}'; fi ); done && \\"
```
i put a sudo in your sudo so you could sudo :D

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -81,6 +147,9 @@ failure_exit()
     echo "    unlink '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
+    echo "    for loop_device in /dev/loop[0-9]; do ( if grep '${loop_device}' '/proc/mounts'; \\ "
+    echo "    then sudo umount '${rootfs_dir}${loop_device}'; fi ); done && \\"
+    echo "    for loop_device in /dev/loop[0-9]; do unlink '${rootfs_dir}${loop_device}'; done && \\"
```
why two loops?
```sh
    echo "    for loop_device in /dev/loop*; do \\"
    echo "        if grep '\${loop_device}' '/proc/mounts'; then \\"
    echo "            umount '${rootfs_dir}\${loop_device}'; \\"
    echo "        fi; \\"
    echo "        if [ -f '${rootfs_dir}\${loop_device}' ]; then \\"
    echo "            unlink '${rootfs_dir}\${loop_device}' ]; \\"
    echo "        fi; \\"
    echo "  done && \\"
```
you want to escape the $ on loop_device, to prevent it from resolving, but ${rootfs_dir} should be resolved.

more importantly however:
loop[0-9] won't umount loop10
I am quite unhappy with the /dev/loop* however as it will umount ALL loop devices, INCLUDING the ones I have unrelated to this test!
I think what would be better, is to get all loop mount devices from /proc/mount that match rootfs_dir.
something like
```sh
loop_devices="$(grep '"${rootfs_dir}"*/dev/loop' /proc/mounts)"
```
Not sure if the * works without special quotes or anything, and not sure if you can do this with busybox grep. I think you need grep -E (to do regex) and then the asterix becomes .* (e.g. match any character inbetween) (if the order is even right). But that will atleast all valid devices!

see https://www.thegeekstuff.com/2011/10/grep-or-and-not-operators for reference

raysiudak replied:
Yes. 
The code is changed, the scenario where failrue exit is printed and the loop devices are mounted is not possible anymore, therefore removed the this code.

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +228,26 @@ test_system_update_entrypoint()
 
+test_execute_resize_partition_shrink_spool_ok()
+{
+    partition_table_file="/tmp/partition_table"
```
why not use chroot ${rootfs_dir} mktemp? (if you can't store it in a variable instead)
remember however that you also need to clean it up when things fail!

raysiudak replied:
Putting it in a variable was what i had before we decided to do it with the toolbox tools in the chroot -ed environment.    
I don't think it is possible, normally you would 'echo partition_table | fdisk image_file' i don't see how this can be done in chroot.

oliv3r replied:
i'm not seeing the cleanup of the file however now.
Maybe pre-generate a in setup() and clean it up in teardown().

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +228,26 @@ test_system_update_entrypoint()
+
+    spool_size="$(( SECTOR_TOTAL - SPOOL_START ))"
+    new_spool_size="$(( spool_size / 2 ))"
+    new_spool_start="$(( SPOOL_START + new_spool_size ))"
```
while completely relevant here, this is a note to ourselves, the final 'actual' partition table will need to account for partition and file-system alignment to flash erase sizes!

oliv3r commented on tests/buildenv.sh:
```diff
@@ -69,7 +68,7 @@ check_emulation_support()
     fi
 
-    echo "ok"
+    echo "${ARM_EMU_BIN} ok"
```
:+1:

oliv3r commented on tests/buildenv.sh:
```diff
@@ -68,7 +68,7 @@ check_emulation_support()
     fi
 
-    echo "${ARM_EMU_BIN} ok"
+    echo "${ARM_EMU_BIN}: ok"
```
i think this change was part of your other fixup ;)

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -7,13 +7,16 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
-SECTOR_SIZE="512"
-SECTOR_TOTAL="7553024"
+# Test storage device parameters.
+# All storage start and and parameters are sector numbers. actual byte positions
```
and end
instead of
and and

sector numbers. Actual byte

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -7,13 +7,16 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
+# are calculated by multiplying the sector number with the sector size.
+STORAGE_DEVICE_IMG="/tmp/storage_device.img"
+SECTOR_SIZE_B="512"
+STORAGE_END="7553024"
```
To be consistent with the above, call it STORAGE_DEVICE_IMG_END (and sort it next to it)
also a comment  behind it, indicating the true size would be helpful

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -7,13 +7,16 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
+# All storage start and and parameters are sector numbers. actual byte positions
+# are calculated by multiplying the sector number with the sector size.
+STORAGE_DEVICE_IMG="/tmp/storage_device.img"
+SECTOR_SIZE_B="512"
```
what does the B mean here?

besides, why did you change it? I think it was fine before :)

raysiudak replied:
BYTES but ok i will change it back

oliv3r commented on Dockerfile:
```diff
@@ -4,7 +4,12 @@ LABEL Maintainer="software-embedded-platform@ultimaker.com" \
 RUN apk add --no-cache \
+        e2fsprogs \
+        f2fs-tools \
+        sed \
```
i still think this is unnecessary for now as afaik the busybox variant does exactly what you request it to do. but it's okay to leave it in as it makes you happy :)

oliv3r commented on tests/buildenv.sh:
```diff
@@ -5,8 +5,8 @@ set -eu
-FILESYSTEMS="overlay squashfs tmpfs"
-COMMANDS="apk mksquashfs xz"
+FILESYSTEMS="ext4 f2fs overlay squashfs tmpfs"
+COMMANDS="apk mkfs.ext4 mkfs.f2fs mksquashfs sed sfdisk xz"
```
:+1: irregardless of which variant of sed is used, it should be in this list

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,23 @@
+# are calculated by multiplying the sector number with the sector size.
+STORAGE_DEVICE_IMG="/tmp/storage_device.img"
+SECTOR_SIZE="512"
+STORAGE_DEVICE_END="7553024"
```
still missing comment at the end about the size
```sh
STORAGE_DEVICE_END="7553024" # sectors, about 3.6 GiB
```

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,23 @@
+SECTOR_SIZE="512"
+STORAGE_DEVICE_END="7553024"
+BOOT_START="2048"
+ROOTFS_START="67584"
```
```sh
ROOTFS_START="67584" # offset 33 MiB
```
or something like that

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -3,8 +3,23 @@
+STORAGE_DEVICE_END="7553024"
+BOOT_START="2048"
+ROOTFS_START="67584"
+USERDATA_START="1998848"
```
```sh
USERDATA_START="1998848" # offset 976 MiB
```

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -13,6 +28,34 @@ result=0
+
+    sfdisk "${rootfs_dir}${STORAGE_DEVICE_IMG}" << \
+______________________________________________________________________________________
+unit: sectors
```
for completeness sake, i'd add the 'label: dos' marker above the unit as well, to ensure we do not generate GPT partitions (we currently cannot support proper GPT on our platform)

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -40,7 +83,14 @@ setup()
+    create_dummy_storage_device
+
+    sfdisk -d "${LOOP_STORAGE_DEVICE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}"
+
```
one newline too much

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -40,7 +83,14 @@ setup()
     mount --bind /proc "${rootfs_dir}/proc" 1> /dev/null
     ln -s ../proc/self/mounts "${rootfs_dir}/etc/mtab"
 
+    mount -t devtmpfs none "/dev"
```
Is this /dev device to be intended to be abailable in the chrooted environment? or in the host.

If it is supposed to be available in the chroot, then this should be mounted at "${rootfs_dir}/dev"
if it is supposed to be mounted in the host, we have to first check if it is not already mounted, (which is only unlikely if we are in the docker environment)

raysiudak replied:
It is intended to be on the host. Ok will and a check

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -60,6 +122,10 @@ teardown()
     fi
 
+    if grep -q "/dev" /proc/mounts; then
+        umount "/dev" || result=1
```
same here, if it is the chrooted env. then rootfs_dir is missing, otherwise ... tricky how to solve this, as we do not want to umount our hosts /dev. But since the command below does /proc for rootfs_dir, my guesstimate says its missing rootfs_dir?

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -81,6 +147,9 @@ failure_exit()
     echo "    unlink '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
+    echo "    losetup -d '${LOOP_STORAGE_DEVICE}' && \\"
+    echo "    unlink '${STORAGE_DEVICE_IMG}' && \\"
+    echo "    umount '/dev' && \\"
```
as above missing rootfs_dir?

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -107,7 +178,7 @@ run_test()
         fi
     fi
-    printf '\n'
+    echo "______________________________________________________________________________________"
```
:+1:

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -97,6 +166,8 @@ run_test()
     setup
 
+    echo "______________________________________________________________________________________"
+    echo
```
:trophy:  ah, that's even nicer then echo ""

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +230,25 @@ test_system_update_entrypoint()
+test_execute_resize_partition_grow_rootfs_ok()
+{
+    userdata_size="$((STORAGE_DEVICE_END - USERDATA_START))"
+    new_userdata_size="$((userdata_size / 2))"
```
for this test case it won't matter, but always be weary of divisions as the the result will be truncated (integer div only)

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +230,25 @@ test_system_update_entrypoint()
 
+test_execute_resize_partition_grow_rootfs_ok()
+{
+    userdata_size="$((STORAGE_DEVICE_END - USERDATA_START))"
```
don't we need $ markers for the variables? I guess not, i think shell check even says we don't need them! I probably have to remove them in my other script then ;)

raysiudak replied:
Try to shellcheck "$((${STORAGE_DEVICE_END} - ${USERDATA_START}))"

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +230,25 @@ test_system_update_entrypoint()
+    new_userdata_size="$((userdata_size / 2))"
+    new_userdata_start="$((USERDATA_START + new_userdata_size))"
+
+    sed -i "s/p3.*type/p3 : start=     ${new_userdata_start}, size=     ${userdata_size}, type/" \
```
can you add a simple oneliner comment above BRIEFLY explaining the regular expression?
(I think its something like replace partition something?)

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -159,6 +230,25 @@ test_system_update_entrypoint()
+    sed -i "s/p3.*type/p3 : start=     ${new_userdata_start}, size=     ${userdata_size}, type/" \
+        "${rootfs_dir}${PARTITION_TABLE_FILE}"
+
+    sha512sum "${rootfs_dir}${PARTITION_TABLE_FILE}" > "${rootfs_dir}${PARTITION_TABLE_FILE}.sha512" || return 1
```
i kind of started to agree with your use of return 1 as documentation. Maybe if we ever get time to do EMP-374 this will be more apparant as well

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -83,14 +84,15 @@ setup()
 
-    mount -t devtmpfs none "/dev"
+    if ! grep -q "/dev" /proc/mounts; then
+        mount -t devtmpfs none "/dev"
```
can we look at this together? I still think it should be ${rootfs_dir}/dev

oliv3r commented on tests/toolbox_image.sh:
```diff
@@ -232,10 +234,11 @@ test_system_update_entrypoint()
     new_userdata_size="$((userdata_size / 2))"
     new_userdata_start="$((USERDATA_START + new_userdata_size))"
 
+    # In every line look for a string "p3<don't care>type" and replace it.
```
and replace it to ... why? I'm guessing to match the end of the disk? If that is the case, I think there's a special marker you can put in the file 'start until end'. for the CLI interface, you can do sfdisk start,,L iirc (Granted my script won't work with that yet :) ) if that was the intention. i'm fine with it being explicit as well, you decide.

oliv3r commented on test/toolbox_image.sh:
```diff
@@ -81,6 +158,9 @@ failure_exit()
     echo "    unlink '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
+    echo "    losetup -d '${LOOP_STORAGE_DEVICE}' && \\"
+    echo "    unlink '${STORAGE_DEVICE_IMG}' && \\"
+    echo "    umount '/dev' && \\"
```
you forgot to add your umount steps here :)
```sh
echo "    umount '${rootfs_dir}/dev'"
if "${is_dev_setup_mounted}"; then
    echo "    umount "/dev"
fi
```

raysiudak committed:
Fix pkg artifact not found in the test/toolbox_image.sh script.

The same conditional job execution as implemented for the
builds should be there for the test script.

Contributes to: EMP-272 and EMP-308

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>
